gsy_framework/sim_results/electric_blue/volume_timeseries.py

```diff
@@ -0,0 +1,160 @@
+from gsy_framework.forward_markets.forward_profile import (
+    ForwardTradeProfileGenerator)
+from gsy_framework.sim_results.electric_blue.aggregate_results import (
+    ForwardDeviceStats)
+
+FORWARD_MARKET_TYPES = [
+    AvailableMarketTypes.YEAR_FORWARD, AvailableMarketTypes.MONTH_FORWARD,
+    AvailableMarketTypes.WEEK_FORWARD, AvailableMarketTypes.DAY_FORWARD,
+]
+
+# a dictionary showing the start of time slot for each resolution.
+START_OF = {
+    AggregationResolution.RES_1_HOUR: "hour",
+    AggregationResolution.RES_1_WEEK: "week",
+    AggregationResolution.RES_1_MONTH: "month",
+    AggregationResolution.RES_1_YEAR: "year",
+}
+
+
+class AssetVolumeTimeSeries:
+    """This class generates combined volume time series for the whole year for each asset.
+    The result would be like this for a monthly resolution for the year 2020:
+    {
+        DateTime(2020, 1, 1, 0, 0, 0): {
+            DateTime(2020, 1, 1, 0, 0, 0): {
+                "SSP": 1005.4160297565659,
+                "YEAR_FORWARD": {
+                    "bought_kWh": 201.08320595131318,
+                    "sold_kWh": 201.08320595131318,
+                },
+                "MONTH_FORWARD": {
+                    "bought_kWh": 201.08320595131318,
+                    "sold_kWh": 201.08320595131318,
+                },
+                "WEEK_FORWARD": {
+                    "bought_kWh": 32.432775153437625,
+                    "sold_kWh": 32.432775153437625,
+                },
+                "DAY_FORWARD": {
+                    "bought_kWh": 6.486555030687522,
+                    "sold_kWh": 6.486555030687522,
+                },
+            },
+            DateTime(2020, 2, 1, 0, 0, 0): {...},
+            ...
+            DateTime(2020, 12, 1, 0, 0, 0): {...},
+        },
+    }
+    """
+    def __init__(
+            self, asset_uuid: str, asset_capacity: float,
+            resolution: AggregationResolution):
+
+        self.asset_uuid = asset_uuid
+        self.asset_capacity = asset_capacity
+        self.resolution = resolution
+
+        self._asset_volume_time_series_buffer: Dict[DateTime, Dict] = {}
+
+    def add(self, asset_stats: ForwardDeviceStats, market_type: AvailableMarketTypes):
+        """Add asset time series to asset volume time series."""
+
+        self._add_total_energy_bought(asset_stats, market_type)
+        self._add_total_energy_sold(asset_stats, market_type)
+
+    def save_time_series(self):
+        """Save asset volume time series in the DB."""
+        # TODO: should be implemented.
+        raise NotImplementedError()
+
+    def _add_total_energy_bought(
+            self, asset_stats: ForwardDeviceStats, market_type: AvailableMarketTypes):
+        """Add statistics for the total amount of bought energy."""
+        energy_kWh = asset_stats.total_energy_bought
+        if energy_kWh <= 0:
+            return
+        profile = ForwardTradeProfileGenerator(self.asset_capacity).generate_trade_profile(
```

Comment:
Since `self.asset_capacity` is always the same, it looks like we can instantiate the `ForwardTradeProfileGenerator` just once as an instance variable (without repeating it everytime `_add_total_energy_bought()` and `_add_total_energy_sold()` are called).

Reply:
Makes sense. Thanks for pointing it out.